File: drive.py
# from googleapiclient.discovery import build
# from google.oauth2 import service_account
# SCOPES=['https://www.googleapis.com/auth/drive']
# SERVICE_ACCOUNT_FILE='final-375200-108b7e234cc4.json'
# PARENT_FOLDER_ID='1My3r7IyNwSIHHDqpmvh2xweDGdLYVviH'
# def authenticate():
#     creds=service_account.Credentials.from_service_account_file(SERVICE_ACCOUNT_FILE,scopes=SCOPES)
#     return creds
# def upload(file_path):
#     creds=authenticate()
#     service=build('drive','v3',credentials=creds)
#     file_metadata={
#         'name':"Hello",
#         'parents': [PARENT_FOLDER_ID]
#     }
#     file=service.files().create(
#         body=file_metadata,
#         media=file_path
#     ).execute()

# upload('/home/sarvesh/Pictures/vlcsnap-2022-09-17-12h36m39s263.png')


# from googleapiclient.discovery import build
# from google.oauth2 import service_account
# from googleapiclient.http import MediaFileUpload

# SCOPES = ['https://www.googleapis.com/auth/drive']
# SERVICE_ACCOUNT_FILE = 'final-375200-108b7e234cc4.json'
# PARENT_FOLDER_ID = '1My3r7IyNwSIHHDqpmvh2xweDGdLYVviH'


# def authenticate():
#     creds = service_account.Credentials.from_service_account_file(SERVICE_ACCOUNT_FILE, scopes=SCOPES)
#     return creds


# def upload(file_path):
#     creds = authenticate()
#     service = build('drive', 'v3', credentials=creds)
    
#     file_metadata = {
#         'name': "Hello",
#         'parents': [PARENT_FOLDER_ID]
#     }

#     media_body = media_body = MediaFileUpload(file_path, resumable=True)

#     file = service.files().create(
#         body=file_metadata,
#         media_body=media_body
#     ).execute()

# upload('/home/sarvesh/Pictures/vlcsnap-2022-09-17-12h36m39s263.png')



# from googleapiclient.discovery import build
# from google.oauth2 import service_account
# from googleapiclient.http import MediaFileUpload

# SCOPES = ['https://www.googleapis.com/auth/drive']
# SERVICE_ACCOUNT_FILE = 'final-375200-108b7e234cc4.json'
# PARENT_FOLDER_ID = '1My3r7IyNwSIHHDqpmvh2xweDGdLYVviH'

# def authenticate():
#     creds = service_account.Credentials.from_service_account_file(SERVICE_ACCOUNT_FILE, scopes=SCOPES)
#     return creds

# def upload(file_path):
#     creds = authenticate()
#     service = build('drive', 'v3', credentials=creds)
    
#     file_metadata = {
#         'name': "Hello",
#         'parents': [PARENT_FOLDER_ID]
#     }

#     media_body = MediaFileUpload(file_path, resumable=True)

#     file = service.files().create(
#         body=file_metadata,
#         media_body=media_body
#     ).execute()

#     # Retrieve the file ID
#     file_id = file.get('id')

#     # Fetch file details to get the link
#     file_details = service.files().get(fileId=file_id, fields='webViewLink').execute()

#     # Retrieve and print the link
#     file_link = file_details.get('webViewLink')
#     if file_link:
#         print(f'File uploaded successfully. Link: {file_link}')
#     else:
#         print('Unable to retrieve the file link.')

# upload('/home/sarvesh/Pictures/vlcsnap-2022-09-17-12h36m39s263.png')


from googleapiclient.discovery import build
from google.oauth2 import service_account
from googleapiclient.http import MediaFileUpload
import qrcode
SCOPES = ['https://www.googleapis.com/auth/drive']
SERVICE_ACCOUNT_FILE = 'final-375200-108b7e234cc4.json'
PARENT_FOLDER_ID = '1My3r7IyNwSIHHDqpmvh2xweDGdLYVviH'

def qr(data):
    qr = qrcode.QRCode(
    version=1,
    error_correction=qrcode.constants.ERROR_CORRECT_H,
    box_size=10,
    border=4,
    )
    qr.add_data(data)
    qr.make(fit=True)
    img = qr.make_image(fill='black', back_color='white')
    img.save("qrcode.png")
def authenticate():
    creds = service_account.Credentials.from_service_account_file(SERVICE_ACCOUNT_FILE, scopes=SCOPES)
    return creds

def upload(file_path):
    creds = authenticate()
    service = build('drive', 'v3', credentials=creds)

    file_metadata = {
        'name': "manchester by the sea (again)",
        'parents': [PARENT_FOLDER_ID],
    }

    media = MediaFileUpload(file_path, resumable=True)

    file = service.files().create(
        body=file_metadata,
        media_body=media
    ).execute()

    # Set permissions to allow anyone with the link to view
    service.permissions().create(
        fileId=file['id'],
        body={'role': 'reader', 'type': 'anyone'}
    ).execute()

    # Retrieve and print the shareable link
    file_link = f'https://drive.google.com/file/d/{file["id"]}/view'
    print(f'File uploaded successfully. Shareable link: {file_link}')
    qr(file_link)
upload('/home/sarvesh/Pictures/vlcsnap-2022-09-17-12h36m39s263.png')

File: drive.js
// const fs = require('fs');
// const { google }= require('googleapis');

// const apikeys = require('./final-375200-108b7e234cc4.json');
// const SCOPE = ['https://www.googleapis.com/auth/drive'];

// // A Function that can provide access to google drive api
// async function authorize(){
//     const jwtClient = new google.auth.JWT(
//         apikeys.client_email,
//         null,
//         apikeys.private_key,
//         SCOPE
//     );

//     await jwtClient.authorize();

//     return jwtClient;
// }

// // A Function that will upload the desired file to google drive folder
// async function uploadFile(authClient){
//     return new Promise((resolve,rejected)=>{
//         const drive = google.drive({version:'v3',auth:authClient}); 

//         var fileMetaData = {
//             name:'some.png',    
//             parents:['1My3r7IyNwSIHHDqpmvh2xweDGdLYVviH'] // A folder ID to which file will get uploaded
//         }

//         drive.files.create({
//             resource:fileMetaData,
//             media:{
//                 body: fs.createReadStream('some.png'), // files that will get uploaded
//                 mimeType:'image/jpeg'
//             },
//             fields:'id'
//         },function(error,file){
//             if(error){
//                 return rejected(error)
//             }
//             resolve(file);
//         })
//     });
// }

// authorize().then(uploadFile).catch("error",console.error()); // function call


// drive upload success
const fs = require('fs');
const { google } = require('googleapis');

const apikeys = require('./final-375200-108b7e234cc4.json');
const SCOPE = ['https://www.googleapis.com/auth/drive'];

// A Function that can provide access to Google Drive API
async function authorize() {
    const jwtClient = new google.auth.JWT(
        apikeys.client_email,
        null,
        apikeys.private_key,
        SCOPE
    );

    await jwtClient.authorize();

    return jwtClient;
}

// A Function that will upload the desired file to a Google Drive folder
async function uploadFile(authClient) {
    return new Promise((resolve, reject) => {
        const drive = google.drive({ version: 'v3', auth: authClient });

        const fileMetaData = {
            name: 'some.png',
            parents: ['1My3r7IyNwSIHHDqpmvh2xweDGdLYVviH'] // A folder ID to which file will get uploaded
        };

        drive.files.create({
            resource: fileMetaData,
            media: {
                body: fs.createReadStream('some.png'), // Specify the correct file path
                mimeType: 'image/png' // Adjust mimeType if necessary
            },
            fields: 'id'
        }, (error, file) => {
            if (error) {
                return reject(error);
            }
            console.log('Uploaded file object:', file); // Log the file object
            const fileId = file.data.id;
            const fileLink = `https://drive.google.com/file/d/${fileId}/view`;
            console.log('File uploaded successfully. View link:', fileLink);
            resolve(file);
        });
    });
}

authorize().then(uploadFile).catch(console.error);
